tasks/Functions-client/buildOffchainSecrets.js

```diff
@@ -0,0 +1,115 @@
+const { getDecodedResultLog } = require("../../FunctionsSandboxLibrary")
+const { encrypt, encryptWithSignature } = require('../../FunctionsSandboxLibrary/encryptSecrets')
+const { networkConfig } = require('../../network-config')
+const fs = require('fs')
+
+task("functions-build-offchain-secrets", "Builds an off-chain secrets object for one or many nodes that can be uploaded and referenced via URL")
+  .addOptionalParam("output", "Output file name (defaults to offchain-secrets.json)")
+  .setAction(async (taskArgs) => {
+    if (network.name === "hardhat") {
+      throw Error(
+        'This command cannot be used on a local hardhat chain.'
+      )
+    }
+
+    const requestConfig = require("../../Functions-request-config")
+
+    // Verify that perNodeSecrets and/or secrets is correctly specified in the config
+    if (requestConfig.perNodeSecrets && !Array.isArray(requestConfig.perNodeSecrets)) {
+      throw Error('perNodeSecrets is not correctly specified in config file.  It must be an array of objects.')
+    }
+
+    if (requestConfig.secrets && typeof requestConfig.secrets !== 'object') {
+      throw Error('secrets object is not correctly specified in config file.  It must be an object.')
+    }
+
+    if (
+      (!requestConfig.perNodeSecrets || requestConfig.perNodeSecrets.length === 0)
+      && (!requestConfig.secrets || Object.keys(requestConfig.secrets).length === 0)
+    ) {
+      throw Error('Neither perNodeSecrets nor secrets is specified')
+    }
+
+    const defaultSecretsObjectKeys = requestConfig.secrets
+      ? JSON.stringify(Object.keys(requestConfig.secrets).sort())
+      : undefined
+
+    if (!requestConfig.secrets || defaultSecretsObjectKeys.length === 0) {
+      console.log("\nWARNING: No default `secrets` provided.  If DON membership changes, the new node will not be able to process requests.\n")
+    }
+
+    if (requestConfig.perNodeSecrets) {
+      for (const secrets of requestConfig.perNodeSecrets) {
+        if (typeof secrets !== 'object') {
+          throw Error('perNodeSecrets is not correctly specified in config file.  It must be an array of objects.')
+        }
+  
+        if (Object.keys(secrets).length === 0) {
+          throw Error('In the config file, perNodeSecrets contains an empty object.')
+        }
+  
+        const secretsObjectKeys = JSON.stringify(Object.keys(secrets).sort())
+  
+        if (
+          requestConfig.secrets
+          && defaultSecretsObjectKeys.length > 0
+          && defaultSecretsObjectKeys !== secretsObjectKeys
+        ) {
+          throw Error('In the config file, not all objects in `perNodeSecrets` have the same object keys as default `secrets`. (The values can be different, but the keys should be the same between all objects.)')
```

Comment:
I wonder if we could relax that to allow per-node secrets for one auth key and a global default for another. Not important right now